### PR TITLE
Fixing 2D moving platform logic

### LINKED ISSUE #27537: Experimenting with gravity and moving platforms, head gets stuck
**Godot version:**
3.1 Stable Windows Mono

**OS/device including version:**
Windows 10

**Issue description:**
Using MoveAndSlide, I had to set the collision Safe Margin down to .05 to avoid sliding on moving platforms.  But that introduced an issue where I couldn't jump off platforms moving against gravity.  I added code that, during the first physics frame after jumping or flipping gravity, I just do a move without detection.  At all other times, I do a normal MoveAndSlide.  The issue is that when I reverse gravity, and I was floored to an object that is moving torward me, my now head gets stuck in the platform until it reverses direction.

I don't EVER want my head to do Floor detection, so I'm not sure why I'm getting stuck.

**Steps to reproduce:**
Add KinematicBody2D platforms moving up and down with a normal AnimationPlayer.  Add a KinematicBody2D player.  During initial frame after gravity change/jump, set Position instead of MoveAndSlide.  Jumping works, but gravity change gets head stuck.

**Minimal reproduction project:**
[Platform Game.zip](https://github.com/godotengine/godot/files/3025791/Platform.Game.zip)
This is my first day of playing with this, so please be kind.  I made the Sprite, but the background is placeholder from an online tutorial.

*Bugsquad edit:* MRP with gdscript: [Platform.Game.zip](https://github.com/godotengine/godot/files/6766540/Platform.Game.zip)

### LINKED ISSUE #40256: CircleShape2D weird behavior with Moving Platforms
**Godot version:**

3.2.2.stable.official

**OS/device including version:**

Solus 4.1 Gnome 3.36.3

**Issue description:**

Using a CollisonShape2D with a CircleShape2D causes KinematicBody2D to behave weirdly when dealing with moving platforms. Check the minimal reproduction project for implementation details.

Bugs noticed:
- When the platform moves up the Player can't jump
moving-platform-jump-bug

- If `CollisionShape2D.one_way_collision = true` the above bug is fixed but the KinematicBody2D some times passes through the platform
one-way-collision-bug

Both problems are fixed when using a RectangleShape2D

rectangle-fix

**Steps to reproduce:**
1. Implement basic movement using `KinematicBody2D.move_and_slide_with_snap()`
2. Implement moving platforms with `KinematicBody2D.sync_to_physics = true` and an `AnimationPlayer`, optionally set the `AnimationPlayer.playback_process_mode = ANIMATION_PROCESS_PHYSICS`
3. Use `CircleShape2D` on the Player's `KinematiBody2D/CollisionShape2D.shape`
4. Animate the Platform moving up and down
5. Jump on the platform and try to jump on it when it goes up
6. Set the Platform's `KinematicBody2D/CollisionShape2D.one_way_collision = true`
7. Repeat step 5 with quick hops

**Minimal reproduction project:**

[circleshape-bug-minimal-project.zip](https://github.com/godotengine/godot/files/4903142/circleshape-bug-minimal-project.zip)

### LINKED ISSUE #42765: move_and_slide causes unexpected slide on edge of fast moving platforms
**Godot version:** 3.2.3 official & 4.0 (a392aa455f)

**OS/device including version:** macOS 10.15.7 (not related to graphics)

**Issue description:**

`move_and_slide` makes KinematicBody2D slide horizontally on a up-moving platform if the KinematicBody2D is on the edge and the platform is moving fast enough.

demo

In the demo project, only a downward velocity (gravity) is applied to the player. The platform is a KinematicBody2D with sync_to_physics turned on and is moved by an AnimationPlayer. `move_and_slide` makes the player move right when elevating.

* The MRP is using `move_and_slide_with_snap`, but `move_and_slide` behaves the same.
* The MRP is made using 3.2.3. In 4.0, the slow platform speed have to be halved, i.e. the safe speed in 3.2 is bad in 4.0.
* Something I don't understand: If you run the MRP, the player slides off the platform immediately on start. But after clicking "Reset" to `reload_current_scene`, it stays on the platform by default.

I tried some pending PRs fixing issues related to `move_and_slide*`, but they didn't fix this issue. (Issue #28895 seems similar, but its PR fix does not fix this issue.)

**Steps to reproduce:**
Use the MRP, and turn on "Fast Platform".

**Minimal reproduction project:**
[SlideOff.zip](https://github.com/godotengine/godot/files/5371141/SlideOff.zip)

### LINKED ISSUE #47366: Body stuck when a moving platform goes up (especially if the collision is a tilemap)
**Godot version:**
3.3.RC6

**Issue description:**

I opened the following issue while ago #35550 however it was not clear, especially now where there was multiple fix in physics.

When the platform move, especially if the collision is a tilemap, the body is stuck when the platform goes up, and at a greater speed, passed through.

Note In the project/video: the camera follow the platform, the body move constantly  it should never stop.

https://user-images.githubusercontent.com/6397893/112530530-d4d8d600-8da6-11eb-8b4c-5d78876fc12b.mp4

**Minimal reproduction project:**

[movingTilemap.zip](https://github.com/godotengine/godot/files/6207281/movingTilemap.zip)

### LINKED ISSUE #50155: KinematicBody2D stops falling when it comes into contact with the wall of a moving platform
### Godot version

3.3 Stable

### System information

MacOS 11.4

### Issue description

When a moving platform moves against a falling body, the body can sometimes stop falling.

freeze

### Steps to reproduce

Run the project.

### Minimal reproduction project

[bugwall.zip](https://github.com/godotengine/godot/files/6759937/bugwall.zip)

### LINKED ISSUE #50167: Moving platforms do not work with this stop on slope
### Godot version

3.3 Stable

### System information

MacOS 11.4

### Issue description

Moving platforms do not work with this stop on slope:

platform

### Steps to reproduce

Run the project

### Minimal reproduction project

[stop_on_slope.zip](https://github.com/godotengine/godot/files/6760458/stop_on_slope.zip)

### PR DESCRIPTION
**Pull request by fabriceci**
4.0 version of #50166

*Bugsquad edit*: fixes #42765, fixes #40256, fixes #47366, fixes #50155, fixes #50167, fixes #27537

### CODE REVIEW

**Review by pouleyKetchoupp — APPROVED** [review]
Looks great!

File: servers/physics_server_2d.cpp [hunk 1]
```diff
@@ -630,7 +634,7 @@ void PhysicsServer2D::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("body_set_force_integration_callback", "body", "callable", "userdata"), &PhysicsServer2D::body_set_force_integration_callback, DEFVAL(Variant()));
 
-	ClassDB::bind_method(D_METHOD("body_test_motion", "body", "from", "motion", "infinite_inertia", "margin", "result"), &PhysicsServer2D::_body_test_motion, DEFVAL(0.08), DEFVAL(Variant()));
+	ClassDB::bind_method(D_METHOD("body_test_motion", "body", "from", "motion", "infinite_inertia", "margin", "result", "p_exclude_raycast_shapes", "p_exclude"), &PhysicsServer2D::_body_test_motion, DEFVAL(0.08), DEFVAL(Variant()), DEFVAL(true), DEFVAL(Array()));
```
Comment (akien-mga) [inline servers/physics_server_2d.cpp:637]: The `p_` prefix is used in C++ methods but not in the public bindings.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
This has been merged but GitHub didn't register it properly, so the PR still shows as open (and conflicting with itself).

Let's wait a bit to see if GitHub catches up with this apparent oops it had when I merged.

**Pull request merged**